### ISSUE TITLE
getenv_s() not allowing a null destination?

### ISSUE THREAD

**Issue by ulatekh** [rurban/safeclib#109]
I’m in the middle of porting a very large legacy code base from MS Windows to Linux.
I'm using safeclib to port all the C11 Annex K calls as-is.

I just ran into an incompatibility with `getenv_s()`.
The [public documentation](https://en.cppreference.com/w/c/program/getenv) says `dest` can be `nullptr` and `dmax` can be 0, in which case it just sets the value of `*len`. But safeclib's variant returns an error.

My question is, is this a bug in safeclib, or is safeclib operating from a different specification?
The only online [Annex K specification](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2244.htm#dr_430) I could find on `getenv_s()` indicates it should allow a null `dest` and a zero `dmax`.

**Comment by rurban** [comment]
Good catch. That looks like definitely like an oversight. Will fix it asap

**Comment by rurban** [comment]
I might have been confused, because windows disallows a NULL dest.
Now fixed

**Comment by ulatekh** [comment]
Sorry, just looked at your fix. (I had my own hacky fix.)

It's close, but not quite. The length still needs to be returned even if the destination is null.
The "if (dest)" on line 137 of src/os/getenv_s.c needs to be moved in front of the strcpy_s on line 140, like so:

--- a/src/os/getenv_s.c
+++ b/src/os/getenv_s.c
@@ -134,10 +134,10 @@ EXPORT errno_t _getenv_s_chk(size_t *restrict len, char *restrict dest,
             *len = 0;
         handle_error(dest, dmax, "getenv_s: dmax is too small", ESNOSPC);
         return RCNEGATE(ESNOSPC);
-    } else if (dest) {
+    } else {
         if (len)
             *len = len1;
-        strcpy_s(dest, dmax, buf);
+        if (dest) strcpy_s(dest, dmax, buf);
     }
 
     return EOK;

**Comment by rurban** [comment]
Ouch, you are right
